pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,719 @@
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		clientset.PrependReactor("update", "antreacontrollerinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &v1beta1.AntreaControllerInfo{}, errors.New("error updating controller crd")
+		})
+		controller := newControllerMonitor(clientset)
+		stopCh := make(chan struct{})
+		initController(controller, stopCh)
+		defer close(stopCh)
+		controller.controllerMonitor.syncControllerCRD()
+		assert.Nil(t, controller.controllerMonitor.controllerCRD)
+	})
+}
+
+func TestEnqueueNode(t *testing.T) {
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "TestNode",
+			Namespace: "ns1",
+			Labels:    map[string]string{"en": "vm2"},
+		},
+	}
+	clientset := fakeclientset.NewSimpleClientset()
+	controller := newControllerMonitor(clientset)
+	stopCh := make(chan struct{})
+	initController(controller, stopCh)
+	defer close(stopCh)
+	controller.controllerMonitor.enqueueNode(node)
+	expectedkey, _ := keyFunc(node)
+	obj, _ := controller.controllerMonitor.nodeQueue.Get()
+	assert.Equal(t, expectedkey, obj.(string))
+}
+
+func TestEnqueueExternalNode(t *testing.T) {
+	externalNode := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "TestExternalNode",
+			Namespace: "ns1",
+			Labels:    map[string]string{"en": "vm2"},
+		},
+	}
+	clientset := fakeclientset.NewSimpleClientset()
+	controller := newControllerMonitor(clientset)
+	stopCh := make(chan struct{})
+	initController(controller, stopCh)
+	defer close(stopCh)
+	controller.controllerMonitor.enqueueExternalNode(externalNode)
+	expectedkey, _ := keyFunc(externalNode)
+	obj, _ := controller.controllerMonitor.externalNodeQueue.Get()
+	assert.Equal(t, expectedkey, obj.(string))
+}
+
+func TestProcessNextExternalNodeWorkItem(t *testing.T) {
+	clientset := fakeclientset.NewSimpleClientset()
+	controller := newControllerMonitor(clientset)
+	stopCh := make(chan struct{})
+	initController(controller, stopCh)
+	defer close(stopCh)
+	tc := []struct {
+		name          string
+		key           interface{}
+		expectedError string
+		requeueNum    int
+	}{
+		{
+			name:          "Invalid key format",
+			key:           "namespace/name/error",
+			expectedError: "unexpected key format: \"namespace/name/error\"",
+		},
+		{
+			name:       "Invalid key type",
+			key:        1,
+			requeueNum: 0,
+		},
+		{
+			name:       "Valid key",
+			key:        "default/vm2-e8be5",
+			requeueNum: 0,
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			controller.controllerMonitor.externalNodeQueue.Add(tt.key)
+			controller.controllerMonitor.processNextExternalNodeWorkItem()
+			if tt.expectedError != "" {
+				_, _, err := splitKeyFunc(tt.key.(string))
+				assert.Equal(t, tt.expectedError, err.Error())
+			} else {
+				assert.Equal(t, tt.requeueNum, controller.controllerMonitor.externalNodeQueue.NumRequeues(tt.key))
+			}
+		})
+	}
```

Comment:
I am not sure the test makes sense any more. `requeueNum` is also always 0.
The only reason to keep this test IMO is if you can find a way to have `c.externalNodeLister.ExternalNodes(namespace).Get(name)` return an error (different from NotFound). Can you do it with a reactor like in other tests? I'm not sure this is possible though, maybe @tnqn knows.

Reply:
I don' think a reactor can be added c.externalNodeLister.Should I keep these tests ,Since these are not covered by other tests ...coverage will decrease

Reply:
If a test is not useful, it's better to remove it. Otherwise, it's just code bloat.

Reply:
Removed these tests